syntax = "proto3";

import "google/protobuf/timestamp.proto";

service ProductListService {
    rpc GetAll(Empty) returns (ProductList) {}
    rpc Get(ProductGetRequest) returns (Product) {}
    rpc Add(ProductAddRequest) returns (Product) {}
    rpc Update(ProductUpdateRequest) returns (Product) {}
    rpc Delete(ProductDeleteRequest) returns (Empty) {}
}

message Empty {}

message Product {
    int32 id = 1;
    string text = 2;
    int32 quantity = 3;
    google.protobuf.Timestamp created_date = 4;
    bool completed = 5;
}

message ProductList {
    repeated Product products = 1;
}

message ProductGetRequest {
    int32 id = 1;
}

message ProductDeleteRequest {
    int32 id = 1;
}

message ProductAddRequest {
    string text = 1;
    int32 quantity = 2;
    bool completed = 3;
}

message ProductUpdateRequest {
    int32 id = 1;
    string text = 2;
    int32 quantity = 3;
    google.protobuf.Timestamp created_date = 4;
    bool completed = 5;
}